File: python/src/agents/compliance_check.py
"""
合规检查Agent — 多Agent流水线第三步。

职责：
1. 基于规则引擎检查法律合规性
2. 与标准合同模板进行条款比对
3. 检测缺失的必要条款
4. 输出合规检查报告

检查维度：
- 《中华人民共和国民法典》合同编
- 《中华人民共和国劳动合同法》
- 行业监管要求
"""

from __future__ import annotations

import json
import logging

from langchain_core.messages import HumanMessage, SystemMessage

from ..config import get_llm
from ..pipeline.state import (
    Clause,
    ClauseCategory,
    ComplianceFinding,
    ComplianceStatus,
    ContractReviewState,
)

logger = logging.getLogger(__name__)

SYSTEM_PROMPT = """你是一名中国法律合规检查专家。请对合同条款进行合规性检查。

检查依据：
1. 《中华人民共和国民法典》合同编
2. 《中华人民共和国劳动合同法》（如适用）
3. 行业监管要求
4. 合同基本要素完整性

请输出严格的JSON格式（不要输出其他内容）：

{
  "findings": [
    {
      "clause_id": "条款ID",
      "status": "compliant/non_compliant/needs_review",
      "regulation": "相关法规条文",
      "issue": "合规问题描述",
      "recommendation": "改进建议"
    }
  ],
  "missing_clauses": ["缺失的必要条款名称列表"],
  "overall_compliance": "compliant/non_compliant/needs_review"
}

合规状态说明：
- compliant: 符合法律法规要求
- non_compliant: 违反法律法规
- needs_review: 需要进一步审查"""


# ── 合规规则定义 ──────────────────────────────────────────────

REQUIRED_CLAUSES = {
    "general": [
        "当事人信息",
        "标的/服务内容",
        "价款/报酬",
        "履行期限",
        "违约责任",
        "争议解决",
    ],
    "labor": [
        "劳动合同期限",
        "工作内容和工作地点",
        "工作时间和休息休假",
        "劳动报酬",
        "社会保险",
        "劳动保护",
    ],
    "sale": [
        "标的物描述",
        "数量",
        "价款",
        "交付方式",
        "质量标准",
        "验收条款",
    ],
}

COMPLIANCE_RULES = [
    {
        "name": "invalid_exclusion",
        "keywords": ["免除己方责任", "排除对方主要权利", "加重对方责任"],
        "regulation": "《民法典》第497条",
        "issue": "格式条款中免除己方责任、排除对方主要权利的条款无效",
    },
    {
        "name": "oral_modification",
        "keywords": ["口头变更有效", "口头修改"],
        "regulation": "《民法典》第490条",
        "issue": "合同变更应采用书面形式，口头变更可能产生争议",
    },
    {
        "name": "excessive_penalty",
        "keywords": ["违约金超过损失的30%", "违约金超过百分之三十"],
        "regulation": "《民法典》第585条",
        "issue": "约定的违约金过分高于造成的损失，当事人可以请求减少",
    },
    {
        "name": "labor_probation",
        "keywords": ["试用期超过6个月", "试用期超过六个月"],
        "regulation": "《劳动合同法》第19条",
        "issue": "劳动合同期限三年以上的，试用期不得超过六个月",
    },
]


class ComplianceCheckAgent:
    """合规检查Agent：规则合规检查 + 大模型合规分析。"""

    def __init__(self):
        self.llm = None

    def _ensure_llm(self):
        if self.llm is None:
            self.llm = get_llm()

    def __call__(self, state: ContractReviewState) -> dict:
        """LangGraph节点函数：执行合规检查。"""
        logger.info("合规检查Agent开始处理 review_id=%s", state.review_id)

        if not state.clauses:
            return {
                "compliance_findings": [],
                "overall_compliance": ComplianceStatus.NEEDS_REVIEW,
                "missing_clauses": [],
            }

        try:
            rule_findings = self._rule_based_check(state.clauses)
            missing = self._check_missing_clauses(
                state.clauses, state.contract_type
            )

            self._ensure_llm()
            llm_findings = self._llm_compliance_check(
                state.clauses, state.contract_type
            )

            all_findings = rule_findings + llm_findings
            overall = self._calculate_overall_compliance(all_findings, missing)

            logger.info(
                "合规检查完成：%d个发现，%d个缺失条款，整体合规=%s",
                len(all_findings), len(missing), overall.value,
            )

            return {
                "compliance_findings": all_findings,
                "missing_clauses": missing,
                "overall_compliance": overall,
            }

        except Exception as e:
            logger.error("合规检查Agent异常: %s", str(e))
            rule_findings = self._rule_based_check(state.clauses)
            missing = self._check_missing_clauses(state.clauses, state.contract_type)
            return {
                "compliance_findings": rule_findings,
                "missing_clauses": missing,
                "overall_compliance": ComplianceStatus.NEEDS_REVIEW,
                "errors": state.errors + [f"合规检查LLM分析失败: {str(e)}"],
            }

    def _rule_based_check(self, clauses: list[Clause]) -> list[ComplianceFinding]:
        """基于规则引擎的合规检查。"""
        findings = []
        for clause in clauses:
            for rule in COMPLIANCE_RULES:
                for keyword in rule["keywords"]:
                    if keyword in clause.content:
                        findings.append(
                            ComplianceFinding(
                                clause_id=clause.id,
                                status=ComplianceStatus.NON_COMPLIANT,
                                regulation=rule["regulation"],
                                issue=rule["issue"],
                                recommendation=f"建议修改条款「{clause.title}」中的相关表述",
                            )
                        )
                        break
        return findings

    def _check_missing_clauses(
        self, clauses: list[Clause], contract_type: str
    ) -> list[str]:
        """检查是否缺少必要条款。"""
        required = REQUIRED_CLAUSES.get("general", [])

        ct = contract_type.lower()
        if "劳动" in ct or "labor" in ct:
            required = required + REQUIRED_CLAUSES.get("labor", [])
        elif "买卖" in ct or "sale" in ct or "采购" in ct:
            required = required + REQUIRED_CLAUSES.get("sale", [])

        existing_titles = " ".join(c.title + " " + c.content[:50] for c in clauses)
        missing = []

        category_map = {
            "当事人信息": ["甲方", "乙方", "party", "当事人"],
            "标的/服务内容": ["标的", "服务内容", "工作内容", "商品"],
            "价款/报酬": ["价款", "报酬", "费用", "金额", "价格"],
            "履行期限": ["期限", "交付", "完成时间", "交货"],
            "违约责任": ["违约", "责任", "赔偿"],
            "争议解决": ["争议", "仲裁", "诉讼", "管辖"],
        }

        for req in required:
            keywords = category_map.get(req, [req])
            found = any(kw in existing_titles for kw in keywords)
            if not found:
                missing.append(req)

        return missing

    def _llm_compliance_check(
        self, clauses: list[Clause], contract_type: str
    ) -> list[ComplianceFinding]:
        """大模型合规检查。"""
        clauses_text = "\n\n".join(
            f"[条款ID: {c.id}] {c.title}\n{c.content}" for c in clauses
        )

        messages = [
            SystemMessage(content=SYSTEM_PROMPT),
            HumanMessage(
                content=(
                    f"合同类型：{contract_type}\n\n"
                    f"请检查以下条款的合规性：\n\n{clauses_text}"
                )
            ),
        ]
        response = self.llm.invoke(messages)
        content = response.content.strip()

        if content.startswith("```"):
            content = content.split("\n", 1)[1]
            if content.endswith("```"):
                content = content[:-3]

        result = json.loads(content)
        findings = []
        for item in result.get("findings", []):
            try:
                status = ComplianceStatus(item.get("status", "needs_review"))
            except ValueError:
                status = ComplianceStatus.NEEDS_REVIEW
            if status != ComplianceStatus.COMPLIANT:
                findings.append(
                    ComplianceFinding(
                        clause_id=item.get("clause_id", ""),
                        status=status,
                        regulation=item.get("regulation", ""),
                        issue=item.get("issue", ""),
                        recommendation=item.get("recommendation", ""),
                    )
                )
        return findings

    def _calculate_overall_compliance(
        self,
        findings: list[ComplianceFinding],
        missing: list[str],
    ) -> ComplianceStatus:
        if any(f.status == ComplianceStatus.NON_COMPLIANT for f in findings):
            return ComplianceStatus.NON_COMPLIANT
        if missing or any(f.status == ComplianceStatus.NEEDS_REVIEW for f in findings):
            return ComplianceStatus.NEEDS_REVIEW
        return ComplianceStatus.COMPLIANT

File: golang/internal/agent/compliance_check.go
package agent

import (
	"encoding/json"
	"fmt"
	"log"
	"strings"

	"github.com/bcefghj/multi-agent-contract-review/golang/internal/config"
	"github.com/bcefghj/multi-agent-contract-review/golang/internal/model"
)

type ComplianceCheckAgent struct {
	llmClient *config.LLMClient
}

func NewComplianceCheckAgent(client *config.LLMClient) *ComplianceCheckAgent {
	return &ComplianceCheckAgent{llmClient: client}
}

func (a *ComplianceCheckAgent) Name() string { return "ComplianceCheckAgent" }

func (a *ComplianceCheckAgent) Process(state *model.ReviewState) error {
	log.Printf("[%s] 开始处理 reviewId=%s", a.Name(), state.ReviewID)

	if len(state.Clauses) == 0 {
		state.OverallCompliance = model.NeedsReview
		return nil
	}

	state.MissingClauses = checkMissingClauses(state.Clauses)

	findings, err := a.llmComplianceCheck(state.Clauses, state.ContractType)
	if err != nil {
		log.Printf("[%s] LLM分析失败: %v", a.Name(), err)
		state.Errors = append(state.Errors, fmt.Sprintf("合规检查失败: %v", err))
		state.OverallCompliance = model.NeedsReview
		return nil
	}

	state.ComplianceFindings = findings
	state.OverallCompliance = calcOverallCompliance(findings, state.MissingClauses)

	log.Printf("[%s] 完成：%d个发现，%d个缺失", a.Name(), len(findings), len(state.MissingClauses))
	return nil
}

func checkMissingClauses(clauses []model.Clause) []string {
	required := map[string][]string{
		"当事人信息":   {"甲方", "乙方", "当事人"},
		"标的/服务内容": {"标的", "服务", "商品"},
		"价款/报酬":   {"价款", "报酬", "费用", "金额"},
		"履行期限":    {"期限", "交付", "完成"},
		"违约责任":    {"违约", "责任", "赔偿"},
		"争议解决":    {"争议", "仲裁", "诉讼"},
	}

	allText := ""
	for _, c := range clauses {
		content := c.Content
		if len(content) > 50 {
			content = content[:50]
		}
		allText += c.Title + " " + content + " "
	}

	var missing []string
	for name, keywords := range required {
		found := false
		for _, kw := range keywords {
			if strings.Contains(allText, kw) {
				found = true
				break
			}
		}
		if !found {
			missing = append(missing, name)
		}
	}
	return missing
}

func (a *ComplianceCheckAgent) llmComplianceCheck(clauses []model.Clause, contractType string) ([]model.ComplianceFinding, error) {
	text := ""
	for _, c := range clauses {
		text += fmt.Sprintf("[%s] %s\n%s\n\n", c.ID, c.Title, c.Content)
	}

	systemPrompt := `你是中国法律合规检查专家。输出JSON：
{"findings":[{"clause_id":"","status":"compliant/non_compliant/needs_review","regulation":"","issue":"","recommendation":""}]}`

	response, err := a.llmClient.Chat(systemPrompt, fmt.Sprintf("合同类型：%s\n\n%s", contractType, text))
	if err != nil {
		return nil, err
	}

	jsonStr := extractJSON(response)
	var result struct {
		Findings []struct {
			ClauseID       string `json:"clause_id"`
			Status         string `json:"status"`
			Regulation     string `json:"regulation"`
			Issue          string `json:"issue"`
			Recommendation string `json:"recommendation"`
		} `json:"findings"`
	}

	if err := json.Unmarshal([]byte(jsonStr), &result); err != nil {
		return nil, fmt.Errorf("JSON解析失败: %w", err)
	}

	var findings []model.ComplianceFinding
	for _, f := range result.Findings {
		status := model.ComplianceStatus(f.Status)
		if status != model.Compliant {
			findings = append(findings, model.ComplianceFinding{
				ClauseID:       f.ClauseID,
				Status:         status,
				Regulation:     f.Regulation,
				Issue:          f.Issue,
				Recommendation: f.Recommendation,
			})
		}
	}
	return findings, nil
}

func calcOverallCompliance(findings []model.ComplianceFinding, missing []string) model.ComplianceStatus {
	for _, f := range findings {
		if f.Status == model.NonCompliant {
			return model.NonCompliant
		}
	}
	if len(missing) > 0 {
		return model.NeedsReview
	}
	for _, f := range findings {
		if f.Status == model.NeedsReview {
			return model.NeedsReview
		}
	}
	return model.Compliant
}
